Set storage class of julia globals to dllimport on windows to avoid auto-import weirdness

gbaraldi opened the pull request:
Fixes https://github.com/JuliaLang/julia/issues/54550 on 1.10. Master will require a similar but different PR.

fingolfin commented:
Just to clarify: This PR targets the backports-release-1.10 branch. Does that means 1.11 and 1.12 are not affected? Or will they get separate fixes?

gbaraldi commented:
This is mostly for testing on 1.10 but it's likely this need a similar fix on 1.11 and up

topolarity commented on src/codegen.cpp:
```diff
@@ -493,9 +493,12 @@ struct JuliaVariable {
                 isconst, GlobalVariable::ExternalLinkage,
                 NULL, name);
+        if (Triple(m->getTargetTriple()).isOSWindows())
+            var->setDLLStorageClass(GlobalValue::DLLStorageClassTypes::DLLImportStorageClass); // This is necessary to avoid auto import issues
```
```suggestion
            var->setDLLStorageClass(GlobalValue::DLLStorageClassTypes::DLLImportStorageClass); // Cross-library imports must be explicit for COFF (Windows)
```

topolarity reviewed (approved):
LGTM! A few style nits, mostly.

I can confirm this resolves https://github.com/JuliaLang/julia/issues/54550 for me locally 🎉

topolarity commented on src/codegen.cpp:
```diff
@@ -1786,9 +1789,6 @@ static inline GlobalVariable *prepare_global_in(Module *M, GlobalVariable *G)
         proto->copyAttributesFrom(G);
-        // DLLImport only needs to be set for the shadow module
-        // it just gets annoying in the JIT
-        proto->setDLLStorageClass(GlobalValue::DefaultStorageClass);
```
@vtjnash It'd be good to get a confirmation that this deletion is OK

Keno reviewed (approved):
In the master version of this, you should also be able to turn on the sysimage building test on windows now.

gbaraldi commented on base/linking.jl:
```diff
@@ -110,7 +110,7 @@ function ld()
         flavor = "gnu"
         m = Sys.ARCH == :x86_64 ? "i386pep" : "i386pe"
-        default_args = `-m $m -Bdynamic --enable-auto-image-base --allow-multiple-definition`
+        default_args = `-m $m -Bdynamic --enable-auto-image-base --allow-multiple-definition --disable-auto-import --disable-runtime-pseudo-reloc`
```
We are the only ones

gbaraldi commented on sysimage.mk:
```diff
@@ -16,7 +16,7 @@ $(build_private_libdir)/%.$(SHLIB_EXT): $(build_private_libdir)/%-o.a
 		$(WHOLE_ARCHIVE) $< $(NO_WHOLE_ARCHIVE) \
 		$(if $(findstring -debug,$(notdir $@)),-ljulia-internal-debug -ljulia-debug,-ljulia-internal -ljulia) \
-		$$([ $(OS) = WINNT ] && echo '' -lssp))
+		$$([ $(OS) = WINNT ] && echo '' -lssp --disable-auto-import --disable-runtime-pseudo-reloc))
```
This is already set on the cli, so the executable doesn't have it anyway, this just prevents silent failure like before

Keno commented on src/aotcompile.cpp:
```diff
@@ -1681,16 +1684,16 @@ void jl_dump_native_impl(void *native_code,
+        llvm::IRBuilder<> builder(BasicBlock::Create(Context, "top", F));
+        builder.CreateRet(jlRTLD_DEFAULT_var);
+        F->setLinkage(GlobalValue::ExternalLinkage);
+        if (TheTriple.isOSWindows()) {
```
BinCOFF, rather than windows? ELF on windows does not use these.

gbaraldi commented on src/staticdata.c:
```diff
@@ -2838,9 +2838,9 @@ JL_DLLEXPORT void jl_preload_sysimg_so(const char *fname)
-    void* *jl_RTLD_DEFAULT_handle_pointer;
-    int symbol_found = jl_dlsym(handle, "jl_RTLD_DEFAULT_handle_pointer", (void **)&jl_RTLD_DEFAULT_handle_pointer, 0);
-    if (!symbol_found || (void*)&jl_RTLD_DEFAULT_handle != *jl_RTLD_DEFAULT_handle_pointer)
+    void** (*jl_RTLD_DEFAULT_handle_pointer)(void) = NULL;
+    int symbol_found = jl_dlsym(handle, "get_jl_RTLD_DEFAULT_handle_pointer", (void **)&jl_RTLD_DEFAULT_handle_pointer, 0);
+    if (!symbol_found || (void*)&jl_RTLD_DEFAULT_handle != (jl_RTLD_DEFAULT_handle_pointer()))
```
Having pointer there is correct, this is not a handle but the address which is why it's named differently

giordano commented:
What happened to the history of this PR?

gbaraldi commented:
I believe it was the force pushes to the base branch that this targets?

KristofferC commented:
Hm, didn't mean to cause that. There were a few commits on the backport branch that were erroneous.

gbaraldi committed:
Set storage class of julia globals to dllimport on windows to avoid auto-import weirdness.
Also make the compiler error instead of emitting pseudo-relocations when encountering a reference to a non-exported symbol in a dll.

Co-authored-by: Keno Fischer <keno@juliacomputing.com>
Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>
